test/test_log_forwarding.py

```diff
@@ -0,0 +1,134 @@
+)
+
+
+class TestLogForwarding(LambdaEnvTestCase):
+
+    @property
+    def log_bucket(self) -> str:
+        return aws.qualified_bucket_name(config.logs_term)
+
+    @property
+    def log_file_key(self) -> str:
+        prefix = config.alb_access_log_path_prefix('gitlab')
+        return '/'.join([prefix, 'AWSLogs', '123123123123', 'elasticloadbalancing', '2023', '01', '01', 'test.log.gz'])
+
+    @cached_property
+    def storage_service(self) -> StorageService:
+        return StorageService(bucket_name=self.log_bucket)
+
+    @mock_s3
+    def test(self):
+        self.storage_service.create_bucket()
+        event = chalice.app.S3Event(context={}, event_dict={
+            'Records': [{
+                's3': {
+                    'bucket': {'name': self.log_bucket},
+                    'object': {'key': self.log_file_key}
+                }
+            }]
+        })
+
+        # Test that troublesome characters appear verbatim in the output
+        for chars in (
+            '\\',
+            '\\x22',
+            '%20',
+        ):
+            with self.subTest(chars=chars):
+                input = [
+                    f'https 2022-12-31T23:55:00.388951Z app/azul-gitlab-alb/c051f98624e68d7e 172.71'
+                    f'.0.215:36056 172.71.0.215:80 0.000 0.002 0.000 204 204 963 229 "POST https://'
+                    f'gitlab.dev.singlecell.gi.ucsc.edu:443/api/v4/jobs/request?chars={chars} HTTP/'
+                    f'1.1" "gitlab-runner 15.6.1 (15-6-stable; go1.18.8; linux/amd64)" ECDHE-RSA-AE'
+                    f'S128-GCM-SHA256 TLSv1.2 arn:aws:elasticloadbalancing:us-east-1:122796619775:t'
+                    f'argetgroup/azul-gitlab-http/136c2d6db59941f6 "Root=1-63b0cbd4-7d218b827862950'
+                    f'05dbf8b6d" "gitlab.dev.singlecell.gi.ucsc.edu" "arn:aws:acm:us-east-1:1227966'
+                    f'19775:certificate/81241b8e-c875-4a22-a30e-58003ee139ae" 0 2022-12-31T23:55:00'
+                    f'.386000Z "forward" "-" "-" "172.71.0.215:80" "204" "-" "-"'
+                ]
+                expected_output = [{
+                    'actions_executed': 'forward',
+                    'chosen_cert_arn': 'arn:aws:acm:us-east-1:122796619775:certificate/81241b8e-c87'
+                                       '5-4a22-a30e-58003ee139ae',
+                    'client': '172.71.0.215',
+                    'client:port': '172.71.0.215:36056',
+                    'client_port': 36056,
+                    'domain_name': 'gitlab.dev.singlecell.gi.ucsc.edu',
+                    'elb': 'app/azul-gitlab-alb/c051f98624e68d7e',
+                    'elb_status_code': '204',
+                    'error_reason': '-',
+                    'matched_rule_priority': '0',
+                    'received_bytes': '963',
+                    'redirect_url': '-',
+                    'request': f'POST https://gitlab.dev.singlecell.gi.ucsc.edu:443/api/v4/jobs/req'
+                               f'uest?chars={chars} HTTP/1.1',
```

Comment:
For legibility and maintainability, wrap at meaningful boundaries, like before the path, or at the spaces.